EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs

```diff
@@ -41,34 +46,86 @@ public virtual void ProcessModelFinalizing(IConventionModelBuilder modelBuilder,
                 var typeMapping = (RelationalTypeMapping?)property.FindTypeMapping()
                     ?? _typeMappingSource.FindMapping((IProperty)property);
                 var propertyType = Nullable.GetUnderlyingType(property.ClrType) ?? property.ClrType;
-                if (propertyType.IsEnum
-                    && typeMapping != null
-                    && !propertyType.IsDefined(typeof(FlagsAttribute), true)
-                    && property.GetColumnName(tableIdentifier) is { } columnName)
+                if (!propertyType.IsEnum
+                    || typeMapping is null
+                    || propertyType.IsDefined(typeof(FlagsAttribute), true)
+                    || property.GetColumnName(tableIdentifier) is not { } columnName)
                 {
-                    var enumValues = Enum.GetValues(propertyType);
-                    if (enumValues.Length <= 0)
-                    {
-                        continue;
-                    }
+                    continue;
+                }
+
+                var enumValues = Enum.GetValues(propertyType).Cast<object>().Select(typeMapping.GenerateSqlLiteral).ToList();
+                if (enumValues.Count == 0)
+                {
+                    continue;
+                }
 
-                    sql.Clear();
+                sql.Clear();
 
-                    sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+                sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+
+                if (TryParseMinAndMax(enumValues, out var minValue, out var maxValue))
+                {
+                    sql.Append(" BETWEEN ");
+                    sql.Append(minValue);
+                    sql.Append(" AND ");
+                    sql.Append(maxValue);
+                }
+                else
+                {
                     sql.Append(" IN (");
                     foreach (var item in enumValues)
                     {
-                        var value = typeMapping.GenerateSqlLiteral(item);
-                        sql.Append($"{value}, ");
+                        sql.Append(item);
+                        sql.Append(", ");
                     }
 
                     sql.Remove(sql.Length - 2, 2);
-                    sql.Append(")");
-
-                    var constraintName = $"CK_{tableName}_{columnName}_Enum";
-                    entityType.AddCheckConstraint(constraintName, sql.ToString());
+                    sql.Append(')');
                 }
+
+                var constraintName = $"CK_{tableName}_{columnName}_Enum";
+                entityType.AddCheckConstraint(constraintName, sql.ToString());
             }
         }
     }
+
+    // we use decimal explicitly here because decimal is a wider number type than all of the valid enum backing types
+    private bool TryParseMinAndMax(IReadOnlyCollection<string> enumValues, out decimal? minValue, out decimal? maxValue)
```

Comment:
I thought about this again, specifically in the context of https://github.com/efcore/EFCore.CheckConstraints/pull/96#discussion_r1066853032... It's not great for us to be dealing with regex and potentially database-specific pattern matching and parsing on the string literals. For example, at least in theory some database may have a locale-sensitive numeric representation, where we'd have commas or dots in the string representation. So here are some thoughts on an alternate approach (am open to discussing your approach too though).

First, for the value converter case (https://github.com/efcore/EFCore.CheckConstraints/pull/96#discussion_r1067162850), it's possible to get the "provider CLR type" (that is, the CLR type after any value converter did its thing) by accessing `typeMapping.Converter.ProviderClrType`; this can tell us simply and directly whether the enum is mapped to a numeric column in the database, regardless of the model CLR type (which is an enum).

At that point, we can run the value converter over all the model enum values to get all the converted provider numeric values (as opposed to a string literal representation which needs to be parsed back), and then simply work on that (sort, min, max). Of course, it there's no value converter we can just use the model enum type's underlying values.

What do you think?

Reply:
Ah, great! Yeah that would help a lot! I'll play around with it, should greatly simplify the checks.

Reply:
@roji Thanks again for the tip - the new code is much simpler than before.

One question, I see someone requested the ability to generate checks for multiple separate ranges. Is that something you'd like for this initial PR?

Reply:
That seems a bit edge-casey to me, and it might even be inefficient compared to just a single IN clause...

I'd suggest, let's deliver the simpler version first and see what people want after...

Reply:
👍